zipline/utils/memoize.py

```diff
@@ -84,4 +87,194 @@ def __get__(self, instance, owner):
+                with lock:
+                    try:
+                        result = cache[key]
+                        cache_renew(key)    # record recent use of this key
+                        hits[0] += 1
+                        return result
+                    except KeyError:
+                        pass
+                result = user_function(*args, **kwds)
+                with lock:
+                    cache[key] = result     # record recent use of this key
+                    misses[0] += 1
+                    if len(cache) > maxsize:
+                        # purge least recently used cache entry
+                        cache_popitem(False)
+                return result
+
+        def cache_info():
+            """Report cache statistics"""
+            with lock:
+                return hits[0], misses[0], maxsize, len(cache)
+
+        def cache_clear():
+            """Clear the cache and cache statistics"""
+            with lock:
+                cache.clear()
+                hits[0] = misses[0] = 0
+
+        wrapper.cache_info = cache_info
+        wrapper.cache_clear = cache_clear
+        return wrapper
+
+    return decorating_function
+
+
+class _WeakArgs(list):
+    """
+    Works with _WeakArgsDict to provide a weak cache for function args.
+    When any of those args are gc'd, the pair is removed from the cache.
+    """
+    def __init__(self, items, dict_remove=None):
+        def remove(k, selfref=ref(self), dict_remove=dict_remove):
+            self = selfref()
+            if self is not None and dict_remove is not None:
+                dict_remove(self)
+
+        super(_WeakArgs, self).__init__(tuple(self._try_ref(item, remove)
+                                              for item in items))
+
+    @staticmethod
+    def _try_ref(item, callback):
+        try:
+            return ref(item, callback)
+        except TypeError:
+            return item
+
+    def __hash__(self):
+        try:
+            return self.__hash
+        except AttributeError:
+            h = self.__hash = hash(tuple(self))
+            return h
+
+
+class _WeakArgsDict(WeakKeyDictionary, object):
+    def __delitem__(self, key):
+        del self.data[_WeakArgs(key)]
+
+    def __getitem__(self, key):
+        return self.data[_WeakArgs(key)]
+
+    def __repr__(self):
+        return "<%s at %s>" % (type(self), id(self))
+
+    def __setitem__(self, key, value):
+        self.data[_WeakArgs(key, self._remove)] = value
+
+    def __contains__(self, key):
+        try:
+            wr = _WeakArgs(key)
+        except TypeError:
+            return False
+        return wr in self.data
+
+    def pop(self, key, *args):
+        return self.data.pop(_WeakArgs(key), *args)
+
+
+class _WeakArgsOrderedDict(_WeakArgsDict, object):
+    def __init__(self):
+        super(_WeakArgsOrderedDict, self).__init__()
+        self.data = OrderedDict()
+
+    def popitem(self, last=True):
+        while True:
```

Comment:
Was a vestige of copying from `WeakKeyDictionary.popitem`, where the keys are actual weakrefs, so here we would check if it's still alive.  I think we probably want that behavior ported here instead of missing, in that we should check if all of the weakrefs in the `_WeakArgs` are alive.